### PR TITLE
[BREAKING?] fix(settled, waitUntil): reject with error thrown in run loop

### LINKED ISSUE #310: render does not reject when rendering fails
I would have expected this test to pass, but it fails with "expected not to get here" in addition to an out-of-band qunit failure for the actual `Compiler Error: not-a-component is not a helper`.

```js
test('render can throw', async function(assert) {
  try {
    await render(hbs`{{not-a-component x="y"}}`);
  } catch (err) {
    assert.ok(/not-a-component is not a helper/.test(err.message), "expected to see failure");
  }
  throw new Error("expected not to get here");
});

```

### LINKED ISSUE #453: Only jQuery.click() triggers error caught by assert.throws
This [twiddle reproduces](https://ember-twiddle.com/de9115316563715af87fec8ddec9f796?openFiles=tests.integration.components.my-c-test.js%2Ctemplates.components.my-c.hbs) the issue.

### TLDR
There is a component that will throw an error if two attrs arent passed in (`x` and `y`) and the component is clicked on.

There are three tests:
1. Ensure no errors if both attrs passed in and component is clicked ✅
2. Ensure an error is throw when only one attr is passed in and component is clicked (**uses native JS and tests fails**) ❌
3. Ensure an error is throw when only one attr is passed in and component is clicked (uses jquery and test passes) ✅

Here is where things go wrong: 
```
// Qunit throws
try {
   block.call(currentTest.testEnvironment);  <====== This call doesnt trigger an error when it uses native JS. But FUN FACT: if you highlight this line at a breakpoint, it will log an error (???)
} catch (e) {
   actual = e;
}
...
if (actual) {     <==== `actual` needs to be set otherwise it wont catch the error and the test will fail.
```

_Originally posted by @nlfurniss in https://github.com/emberjs/ember-test-helpers/issues/310#issuecomment-412316982_

### PR DESCRIPTION
**Pull request by buschtoens**
This PR is currently still WIP and failing some tests, but I'd like to receive feedback on the implementation and course-correct, where necessary.

---

The goal of this PR is to fix the longstanding issue of test helpers not propagating run loop errors thrown during their execution, but instead causing an out of band "global failure" that fails the test. For example:

```ts
await render(hbs`<UnknownComponent />`);
// or
await click('.element-throws-on-click');
```

Fixes #310. Fixes #453. Relates to #440, #1128, [ember-qunit#592](https://github.com/emberjs/ember-qunit/issues/592)

---

This PR adds an `rejectOnError` option to `waitUntil`. It defaults to `true`, if the test context is set up correctly for use with `setupOnerror`.

If enabled, `waitUntil` will reject when an error propagates to `Ember.onerror`, using `setupOnerror` under the hood. As such, (a/sync) errors thrown in a run loop will cause `waitUntil` to fail.

Almost all test helpers end with a call to `settled`, which just calls `waitUntil`. Therefore they'll all automatically benefit from this change and reject, if an error is thrown during their execution.

### CODE REVIEW

File: addon-test-support/@ember/test-helpers/wait-until.ts [hunk 1]
```diff
@@ -8,6 +12,17 @@ type Falsy = false | 0 | '' | null | undefined;
...
+  /**
+   * Instrument `Ember.onerror` and reject, when it is called. This is useful
+   * for detecting that an operation inside a run loop has failed.
+   *
+   * This uses {@link setupOnerror}, so it will override any error listeners you
+   * might have set up before.
+   *
+   * @default true if the test context has been setup for usage with {@link setupOnerror}
+   */
+  rejectOnError?: boolean;
```
Comment (buschtoens) [inline addon-test-support/@ember/test-helpers/wait-until.ts:25]: I'm not really fond of this, because:

> it will override any error listeners you might have set up before.

> if the test context has been setup for usage with `setupOnerror`

This limitation arises from how `setupOnerror` works internally. Alternatively, we could _always_ instrument `Ember.onerror` in tests, somewhat like:

```ts
const originalOnError = Ember.onerror;
let onErrorOverride?: typeof Ember.onerror;

Ember.onerror = error => {
  // _Always_ let `waitUntil` know about the error.
  notifyInternalOnError(error);

  // If the user called `setupOnerror(...)`, call the override.
  if (onErrorOverride)
    return onErrorOverride(error);
  
  // Otherwise call the original `Ember.onerror`.
  return onErrorOverride?.(error)
};

export function setupOnerror(onError?: typeof Ember.onerror): void {
  let context = getContext();

  if (!context) {
    throw new Error('Must setup test context before calling setupOnerror');
  }

  if (!cachedOnerror.has(context)) {
    throw new Error(
      '_cacheOriginalOnerror must be called before setupOnerror. Normally, this will happen as part of your test harness.'
    );
  }

  if (typeof onError !== 'function') {
    onError = cachedOnerror.get(context);
  }

  onErrorOverride = onError;
}

export function resetOnerror(): void {
  let context = getContext();
  if (context && cachedOnerror.has(context)) {
    onErrorOverride = cachedOnerror.get(context);
  }
}
```

Comment (buschtoens) [reply addon-test-support/@ember/test-helpers/wait-until.ts:25]: Also, letting `rejectOnError` default to `true` will inevitably break a lot of existing test suites, that currently try to emulate this missing feature by using `setupOnerror` or manually hooking into `Ember.onerror`.

However, I believe that ultimately the behavior implemented in this PR is what a user would expect. So while this breaking change may be a pain, because it will require changing all tests dealing with such erros, I think that it's a _very_ welcome change.

_Maybe_ we could even conjur up a codemod to ease the transition. Or we could add a compatibility mode, that will essentially change the order of `notifyInternalOnError` and `onErrorOverride`, so that "old" tests just continue to work.

```ts
const originalOnError = Ember.onerror;
let onErrorOverride?: typeof Ember.onerror;
let preferOnErrorOverride = true;

Ember.onerror = error => {
  // Compatibility mode for old tests.
  if (preferOnErrorOverride && onErrorOverride)
    return onErrorOverride(error);

  // _Always_ let `waitUntil` know about the error.
  notifyInternalOnError(error);

  // If the user called `setupOnerror(...)`, call the override.
  if (onErrorOverride)
    return onErrorOverride(error);
  
  // Otherwise call the original `Ember.onerror`.
  return onErrorOverride?.(error)
};
```

We should also consider adding a setter trap for `Ember.onerror` to catch users manually overriding it in tests.

**Comment by ef4** [comment]
This seems good. I think it's the best solution that doesn't require changes to ember itself.

But I view this as a workaround for a feature ember itself should offer. There should be public API that lets you implement `render`, and it should include more specific error handling than a single global callback.

**Comment by NullVoxPopuli** [comment]
Should we pursue finishing / deciding what to do with this PR?
